File: js/src/core/reader-view.ts
import { BaseView } from './base-view';
import { ReaderElement, MarkdownPage } from '../types';
import { marked } from 'marked';
import { InvalidParameterError, MarkdownParseError } from '../errors';
import sanitizeHtmlLib from 'sanitize-html';

export interface ReaderContent {
    title: string;
    intro: string;
    elements: ReaderElement[];
}

/**
 * Sanitizes HTML using sanitize-html to prevent XSS attacks
 * More secure than regex-based sanitization with comprehensive protection
 */
const sanitizeHtml = (html: string): string => {
    return sanitizeHtmlLib(html, {
        allowedTags: [
            'p', 'br', 'strong', 'em', 'u', 'i', 'b',
            'h1', 'h2', 'h3', 'h4', 'h5', 'h6',
            'ul', 'ol', 'li',
            'a', 'blockquote', 'code', 'pre',
            'table', 'thead', 'tbody', 'tr', 'th', 'td',
            'hr', 'div', 'span'
        ],
        allowedAttributes: {
            'a': ['href', 'title'],
            '*': ['class', 'id']
        },
        allowedSchemes: ['http', 'https', 'mailto'],
        allowedSchemesByTag: {
            'a': ['http', 'https', 'mailto']
        },
        // Disallow data attributes to prevent data: URL attacks
        allowedClasses: {},
        disallowedTagsMode: 'discard',
        // Enforce strict filtering
        enforceHtmlBoundary: true
    });
};

/**
 * Builds a Reader SGUI view — a scrollable rich-content document.
 *
 * Content elements are appended in order via `addParagraph`, `addSubTitle`,
 * `addImage`, `addMarkdown`, `addListField`, `addLink`, `addTable`,
 * `addCodeBlock`, `addQuote`, `addSeparator` and `addCustomElement`.
 *
 * Extends {@link BaseView}; instantiated by the YeriaApp/YeriaUI factory,
 * populated with these builders, then serialized to a JSON view description and
 * signed into a v3 envelope by `serve()`.
 */
export class ReaderView extends BaseView {

    static fromJson(json: Record<string, unknown>): ReaderView {
        return ReaderView.fromJsonAs(ReaderView, 'Reader', json);
    }
    constructor(viewId: string, title: string, processId?: string) {
        super({
            id: viewId,
            type: 'Reader',
            processId,
            metadata: {
                version: '1.0.0',
                createdAt: new Date()
            }
        });

        this.content = {
            title: title,
            intro: '',
            elements: []
        } as ReaderContent;
    }

    /**
     * Sets the introduction
     */
    setIntro(intro: string): this {
        return this.setIntroText('intro', intro);
    }

    /**
     * Adds a paragraph
     */
    addParagraph(text: string): this {
        const trimmed = text.trim();
        if (trimmed.length === 0) {
            throw new InvalidParameterError('text', text, 'Paragraph text cannot be empty');
        }

        (this.content as ReaderContent).elements.push({
            type: 'paragraph',
            text: trimmed
        });
        return this;
    }

    /**
     * Adds an image
     */
    addImage(url: string, alt?: string, caption?: string): this {
        const trimmedUrl = url.trim();
        if (trimmedUrl.length === 0) {
            throw new InvalidParameterError('url', url, 'Image URL cannot be empty');
        }

        const imageElement: ReaderElement = {
            type: 'image',
            url: trimmedUrl
        };

        if (alt) {
            imageElement.alt = alt;
        }

        if (caption) {
            imageElement.caption = caption;
        }

        (this.content as ReaderContent).elements.push(imageElement);
        return this;
    }

    /**
     * Adds a subtitle
     */
    addSubTitle(text: string): this {
        const trimmed = text.trim();
        if (trimmed.length === 0) {
            throw new InvalidParameterError('text', text, 'Subtitle text cannot be empty');
        }

        (this.content as ReaderContent).elements.push({
            type: 'subtitle',
            text: trimmed
        });
        return this;
    }

    /**
     * Adds Markdown content
     */
    addMarkdown(markdownText: string, options: { sanitize?: boolean } = {}): this {
        const { sanitize = true } = options;

        try {
            const parsed = marked.parse(markdownText);
            const htmlContent = typeof parsed === 'string' ? parsed : parsed.toString();

            const page: MarkdownPage = {
                content: sanitize ? sanitizeHtml(htmlContent) : htmlContent,
                raw: markdownText,
                sanitized: sanitize
            };

            const elements = (this.content as ReaderContent).elements;
            const existingMarkdown = elements.find(
                (element): element is ReaderElement & { type: 'markdown' } => element.type === 'markdown'
            );

            if (!existingMarkdown) {
                elements.push({
                    type: 'markdown',
                    pages: [page]
                });
            } else {
                if (!existingMarkdown.pages) {
                    existingMarkdown.pages = [];
                }

                existingMarkdown.pages.push(page);
            }
        } catch (error) {
            const err = error instanceof Error ? error : new Error('Unknown error');
            throw new MarkdownParseError(this.id, err);
        }

        return this;
    }

    /**
     * Adds a list
     */
    addListField(items: string[], ordered: boolean = false): this {
        if (!Array.isArray(items)) {
            throw new InvalidParameterError('items', items, 'List items must be an array');
        }

        const trimmedItems = items.map(item => item.trim());
        const listElement: ReaderElement = {
            type: 'list',
            items: trimmedItems,
            ordered
        };

        (this.content as ReaderContent).elements.push(listElement);
        return this;
    }

    /**
     * Adds a link
     */
    addLink(url: string, text: string, description?: string): this {
        const trimmedUrl = url.trim();
        const trimmedText = text.trim();

        if (trimmedUrl.length === 0) {
            throw new InvalidParameterError('url', url, 'Link URL cannot be empty');
        }

        if (trimmedText.length === 0) {
            throw new InvalidParameterError('text', text, 'Link text cannot be empty');
        }

        const linkElement: ReaderElement = {
            type: 'link',
            url: trimmedUrl,
            text: trimmedText
        };

        if (description) {
            linkElement.description = description;
        }

        (this.content as ReaderContent).elements.push(linkElement);
        return this;
    }

    /**
     * Adds a table
     */
    addTable(headers: string[], rows: string[][]): this {
        if (!Array.isArray(headers) || !Array.isArray(rows)) {
            throw new InvalidParameterError('table data', { headers, rows }, 'Headers and rows must be arrays');
        }

        const tableElement: ReaderElement = {
            type: 'table',
            headers,
            rows
        };

        (this.content as ReaderContent).elements.push(tableElement);
        return this;
    }

    /**
     * Adds a code block
     */
    addCodeBlock(code: string, language?: string): this {
        const codeElement: ReaderElement = {
            type: 'code',
            code
        };

        if (language) {
            codeElement.language = language;
        }

        (this.content as ReaderContent).elements.push(codeElement);
        return this;
    }

    /**
     * Adds a quote
     */
    addQuote(text: string, author?: string, source?: string): this {
        const quoteElement: ReaderElement = {
            type: 'quote',
            text
        };

        if (author) {
            quoteElement.author = author;
        }

        if (source) {
            quoteElement.source = source;
        }

        (this.content as ReaderContent).elements.push(quoteElement);
        return this;
    }

    /**
     * Adds a separator
     */
    addSeparator(): this {
        (this.content as ReaderContent).elements.push({
            type: 'separator'
        });
        return this;
    }

    /**
     * Adds a custom element
     */
    addCustomElement(type: string, data: Record<string, unknown>): this {
        const trimmedType = type.trim();
        if (trimmedType.length === 0) {
            throw new InvalidParameterError('type', type, 'Custom element type cannot be empty');
        }

        (this.content as ReaderContent).elements.push({
            type: 'custom',
            kind: trimmedType,
            data: { ...data }
        });
        return this;
    }

    /**
     * Gets elements by type
     */
    getElementsByType(type: ReaderElement['type']): ReaderElement[] {
        return (this.content as ReaderContent).elements.filter(element => element.type === type);
    }

    /**
     * Removes an element by index
     */
    removeElement(index: number): boolean {
        const elements = (this.content as ReaderContent).elements;
        if (index >= 0 && index < elements.length) {
            elements.splice(index, 1);
            return true;
        }
        return false;
    }

    /**
     * Inserts an element at a specific index
     */
    insertElement(index: number, element: ReaderElement): boolean {
        const elements = (this.content as ReaderContent).elements;
        if (index >= 0 && index <= elements.length) {
            elements.splice(index, 0, element);
            return true;
        }
        return false;
    }
} 

File: py/yeriasdk/views/reader_view.py
"""
ReaderView - A view for displaying rich content
"""

from typing import Optional, List, Dict, Any
from datetime import datetime
import re

from ..core.base_view import BaseView
from ..types.models import MarkdownPage
from ..errors.exceptions import InvalidParameterError, MarkdownParseError

try:
    import markdown
    MARKDOWN_AVAILABLE = True
except ImportError:
    MARKDOWN_AVAILABLE = False

try:
    import bleach
    BLEACH_AVAILABLE = True
except ImportError:
    BLEACH_AVAILABLE = False

# Tags the renderer accepts (kept in sync with the JS SDK sanitizer).
_ALLOWED_TAGS = [
    "p", "br", "strong", "em", "u", "i", "b",
    "h1", "h2", "h3", "h4", "h5", "h6",
    "ul", "ol", "li",
    "a", "blockquote", "code", "pre",
    "table", "thead", "tbody", "tr", "th", "td",
    "hr", "div", "span",
]
_ALLOWED_PROTOCOLS = ["http", "https", "mailto"]


def _sanitize_html(html_content: str) -> str:
    """Sanitize rendered-markdown HTML, keeping only the allowed tags.

    Uses ``bleach`` (a declared dependency). Disallowed tags are stripped but
    their text is kept, and ``<a>`` hrefs are limited to http/https/mailto —
    the same policy the JS SDK applies, so the wire output is identical
    (allowed tags pass through unescaped, whitespace preserved). The escape
    fallback only runs if bleach is somehow unavailable; escaping is safe (no
    XSS) though it would render tags as literal text.
    """
    if not BLEACH_AVAILABLE:
        import html as html_module
        return html_module.escape(html_content)

    return bleach.clean(
        html_content,
        tags=_ALLOWED_TAGS,
        attributes={"a": ["href"]},
        protocols=_ALLOWED_PROTOCOLS,
        strip=True,
    )


class ReaderView(BaseView):
    """Builds a Reader SGUI view — a scrollable rich-content document.

    Content elements are appended in order via ``add_paragraph``,
    ``add_subtitle``, ``add_image``, ``add_markdown``, ``add_list_field``,
    ``add_link``, ``add_table``, ``add_code_block``, ``add_quote``,
    ``add_separator`` and ``add_custom_element``.

    Extends ``BaseView``; instantiated by the YeriaApp/YeriaUI factory,
    populated with these builders, then serialized to a JSON view description
    and signed into a v3 envelope by ``serve()``.
    """
    @classmethod
    def from_json(cls, json_view):
        """Rehydrate a Reader view from a wire JSON payload."""
        return cls.from_json_as('Reader', json_view)


    def __init__(self, view_id: str, view_title: str, process_id: Optional[str] = None):
        super().__init__(
            {
                "id": view_id,
                "type": "Reader",
                "process_id": process_id,
                "metadata": {
                    "version": "1.0.0",
                    "created_at": datetime.now(),
                },
            }
        )

        self.content = {
            "title": view_title,
            "intro": "",
            "elements": [],
        }

    def set_intro(self, intro: str) -> "ReaderView":
        """Set introduction"""
        return self._set_intro_text("intro", intro)

    def add_paragraph(self, text: str) -> "ReaderView":
        """Add a paragraph"""
        trimmed = text.strip()
        if not trimmed:
            raise InvalidParameterError("text", text, "Paragraph text cannot be empty")

        self.content["elements"].append({"type": "paragraph", "text": trimmed})
        return self

    def add_image(
        self, url: str, alt: Optional[str] = None, caption: Optional[str] = None
    ) -> "ReaderView":
        """Add an image"""
        trimmed_url = url.strip()
        if not trimmed_url:
            raise InvalidParameterError("url", url, "Image URL cannot be empty")

        image_element = {"type": "image", "url": trimmed_url}
        if alt:
            image_element["alt"] = alt
        if caption:
            image_element["caption"] = caption

        self.content["elements"].append(image_element)
        return self

    def add_subtitle(self, text: str) -> "ReaderView":
        """Add a subtitle"""
        trimmed = text.strip()
        if not trimmed:
            raise InvalidParameterError("text", text, "Subtitle text cannot be empty")

        self.content["elements"].append({"type": "subtitle", "text": trimmed})
        return self

    def add_markdown(
        self, markdown_text: str, sanitize: bool = True
    ) -> "ReaderView":
        """Add markdown content"""
        if not MARKDOWN_AVAILABLE:
            raise MarkdownParseError(
                self.id, ValueError("markdown library is required for markdown support")
            )

        try:
            html_content = markdown.markdown(markdown_text)

            page: MarkdownPage = {
                "content": _sanitize_html(html_content) if sanitize else html_content,
                "raw": markdown_text,
                "sanitized": sanitize,
            }

            elements = self.content["elements"]
            existing_markdown = next(
                (e for e in elements if e.get("type") == "markdown"), None
            )

            if not existing_markdown:
                elements.append({"type": "markdown", "pages": [page]})
            else:
                if "pages" not in existing_markdown:
                    existing_markdown["pages"] = []
                existing_markdown["pages"].append(page)

        except Exception as error:
            raise MarkdownParseError(self.id, error) from error

        return self

    def add_list_field(
        self, items: List[str], ordered: bool = False
    ) -> "ReaderView":
        """Add a list"""
        if not isinstance(items, list):
            raise InvalidParameterError(
                "items", items, "List items must be an array"
            )

        trimmed_items = [item.strip() for item in items]
        self.content["elements"].append(
            {"type": "list", "items": trimmed_items, "ordered": ordered}
        )
        return self

    def add_link(
        self, url: str, text: str, description: Optional[str] = None
    ) -> "ReaderView":
        """Add a link"""
        trimmed_url = url.strip()
        trimmed_text = text.strip()

        if not trimmed_url:
            raise InvalidParameterError("url", url, "Link URL cannot be empty")
        if not trimmed_text:
            raise InvalidParameterError("text", text, "Link text cannot be empty")

        link_element = {"type": "link", "url": trimmed_url, "text": trimmed_text}
        if description:
            link_element["description"] = description

        self.content["elements"].append(link_element)
        return self

    def add_table(
        self, headers: List[str], rows: List[List[str]]
    ) -> "ReaderView":
        """Add a table"""
        if not isinstance(headers, list) or not isinstance(rows, list):
            raise InvalidParameterError(
                "table data", {"headers": headers, "rows": rows},
                "Headers and rows must be arrays"
            )

        self.content["elements"].append(
            {"type": "table", "headers": headers, "rows": rows}
        )
        return self

    def add_code_block(
        self, code: str, language: Optional[str] = None
    ) -> "ReaderView":
        """Add a code block"""
        code_element = {"type": "code", "code": code}
        if language:
            code_element["language"] = language

        self.content["elements"].append(code_element)
        return self

    def add_quote(
        self, text: str, author: Optional[str] = None, source: Optional[str] = None
    ) -> "ReaderView":
        """Add a quote"""
        quote_element = {"type": "quote", "text": text}
        if author:
            quote_element["author"] = author
        if source:
            quote_element["source"] = source

        self.content["elements"].append(quote_element)
        return self

    def add_separator(self) -> "ReaderView":
        """Add a separator"""
        self.content["elements"].append({"type": "separator"})
        return self

    def add_custom_element(
        self, type: str, data: Dict[str, Any]
    ) -> "ReaderView":
        """Add a custom element"""
        trimmed_type = type.strip()
        if not trimmed_type:
            raise InvalidParameterError(
                "type", type, "Custom element type cannot be empty"
            )

        self.content["elements"].append(
            {"type": "custom", "kind": trimmed_type, "data": dict(data)}
        )
        return self

    def get_elements_by_type(self, element_type: str) -> List[Dict[str, Any]]:
        """Get elements by type"""
        return [
            e for e in self.content["elements"] if e.get("type") == element_type
        ]

    def remove_element(self, index: int) -> bool:
        """Remove an element by index"""
        elements = self.content["elements"]
        if 0 <= index < len(elements):
            elements.pop(index)
            return True
        return False

    def insert_element(
        self, index: int, element: Dict[str, Any]
    ) -> bool:
        """Insert an element at a specific index"""
        elements = self.content["elements"]
        if 0 <= index <= len(elements):
            elements.insert(index, element)
            return True
        return False

